solutionbox/structured_data/datalab_solutions/structured_data/trainer/util.py

```diff
@@ -140,67 +177,180 @@ def read_examples(input_files, batch_size, shuffle, num_epochs=None):
+  config = tf.contrib.learn.RunConfig(
+      save_checkpoints_secs=args.save_checkpoints_secs)
+
+  train_dir = os.path.join(output_dir, 'train')
+  if args.model_type == 'dnn_regression':
+    estimator = tf.contrib.learn.DNNRegressor(
+        feature_columns=feature_columns,
+        hidden_units=args.layer_sizes,
+        config=config,
+        model_dir=train_dir,
+        optimizer=tf.train.AdamOptimizer(
+          args.learning_rate, epsilon=args.epsilon))
+  elif args.model_type == 'linear_regression':
+    estimator = tf.contrib.learn.LinearRegressor(
+        feature_columns=feature_columns,
+        config=config,
+        model_dir=train_dir,
+        optimizer=tf.train.AdamOptimizer(
+          args.learning_rate, epsilon=args.epsilon))
+  elif args.model_type == 'dnn_classification':
+    estimator = tf.contrib.learn.DNNClassifier(
+        feature_columns=feature_columns,
+        hidden_units=args.layer_sizes,
+        n_classes=train_config['vocab_stats'][target_name]['n_classes'],
+        config=config,
+        model_dir=train_dir,
+        optimizer=tf.train.AdamOptimizer(
+          args.learning_rate, epsilon=args.epsilon))
+  elif args.model_type == 'linear_classification':
+    estimator = tf.contrib.learn.LinearClassifier(
+        feature_columns=feature_columns,
+        n_classes=train_config['vocab_stats'][target_name]['n_classes'],
+        config=config,
+        model_dir=train_dir,
+        optimizer=tf.train.AdamOptimizer(
+          args.learning_rate, epsilon=args.epsilon))
+  else:
+    raise ValueError('bad --model_type value')
+
+  return estimator
+
+
+def preprocess_input(features, target, train_config, preprocess_output_dir, model_type):
+  target_name = train_config['target_column']
+  key_name = train_config['key_column']
+
+  # Do the numerical transforms.
+  with tf.name_scope('numerical_feature_preprocess') as scope:
+    if train_config['numerical_columns'] != []:
+      numerical_analysis_file = os.path.join(preprocess_output_dir,
+                                             NUMERICAL_ANALYSIS)
+      if not ml.util._file.file_exists(numerical_analysis_file):
+        raise ValueError('File %s not found in %s' % 
+                         (NUMERICAL_ANALYSIS, preprocess_output_dir))
+
+      numerical_anlysis = json.loads(
+          ml.util._file.load_file(numerical_analysis_file))
+
+      for name in train_config['numerical_columns']:
+        if name == target_name or name == key_name:
+          continue
+
+        transform_config = train_config['transforms'].get(name, {})
+        transform_name = transform_config.get('transform', None)
+        if transform_name == 'scale':
+          features[name] = _scale_tensor(
+              features[name], 
+              range_min=numerical_anlysis[name]['min'], 
+              range_max=numerical_anlysis[name]['max'], 
+              scale_min=-1, 
+              scale_max=1)
+        elif transform_name == 'max_abs_scale':
```

Comment:
I wonder if we can combine scale and max_abs_scale, and make scale value default to 1?